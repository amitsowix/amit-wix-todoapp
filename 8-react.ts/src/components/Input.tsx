import React, {FC, useState} from 'react';
import {createUseStyles} from 'react-jss'
import {Classes} from 'jss';

interface InputProps {
    onInputSubmit: (text: string) => void;
}

const Input: FC<InputProps> = (props) : JSX.Element => {
    const classes: Classes = useStyles();

    const [inputValue, setInputValue] = useState("");

    const submitInput = (): void => {
        props.onInputSubmit(inputValue)
        setInputValue("");
    }

    return (
        <div className = {classes.inputContainer}>
            <input 
            value = {inputValue}
            autoFocus
            className={classes.input} 
            onChange={(event) => setInputValue(event.target.value)} 
            onSubmit={()=>{
                if (inputValue !== ""){
                    submitInput();
                }}}
            onKeyDown= {(event) => {
                if (event.key === "Enter" && inputValue !== ""){
                    submitInput();
                }}}
            />
            <div className={classes.submitButton} onClick={()=>props.onInputSubmit(inputValue)}>
                GO
            </div>
        </div>
    )            
}

const useStyles: () => Classes = createUseStyles({
    input: {
        marginLeft: "10px",
        outline: "none",
        border: "none",
        borderBottom: "1px solid gray"
    },
    
    submitButton: {
        marginLeft: "10px",
        border: "1px solid gray",
        padding: "6px",
        borderRadius: "5px",
        cursor: "pointer"
    },

    inputContainer: {
        display: "flex"
    }
});

export default Input;