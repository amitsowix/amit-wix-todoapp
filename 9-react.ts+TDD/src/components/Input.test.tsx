import React from 'react';
import renderer, {ReactTestRenderer, ReactTestRendererJSON} from 'react-test-renderer';
import {screen, render} from '@testing-library/react';
import Input from './Input';
import {mount} from 'enzyme';


describe('Test Render', () => {
    test('Component Rendered Correctly', () => {
        const component: ReactTestRenderer = renderer.create(
            <Input name={'test'} onInputSubmit={()=>{}}/>
        )
        let tree: ReactTestRendererJSON | ReactTestRendererJSON[] | null = component.toJSON();
        expect(tree).toMatchSnapshot();
    })

    test('Container type', () => {
        const component: ReactTestRenderer = renderer.create(
            <Input name={'test'} onInputSubmit={()=>{}}/>
        );
        let tree: ReactTestRendererJSON | ReactTestRendererJSON[] | null = component.toJSON();
        if (tree && !Array.isArray(tree)){
            expect(tree.type).toMatch('div');
        }
    })

    test('Submit Button text', () => {
        render(<Input name={'test'} onInputSubmit={()=>{}}/>);
        expect(screen.getByText('GO')).toBeInTheDocument();
    })

    test('Component class', () => {
        const component: ReactTestRenderer = renderer.create(
            <Input name={'test'} onInputSubmit={()=>{}}/>
        );
        let tree: ReactTestRendererJSON | ReactTestRendererJSON[] | null = component.toJSON();
        if (tree && !Array.isArray(tree)){
            expect(tree.props.className).toMatch('inputContainer-0-2-3');
        }
    })
})


describe('Submit functionality test', () => {
    test('Submit Button click with value', () => {
        const mockCallBack = jest.fn();
        const wrapper = mount(<Input name={'test'} onInputSubmit={mockCallBack} />);
        const submitButton = wrapper.find('.submitButton-0-2-2');
        const input = wrapper.find('.input-0-2-1');
        input.simulate('focus');
        input.simulate('change', { target: { value: 'Changed' } });
        submitButton.simulate('click');
        expect(mockCallBack.mock.calls.length).toEqual(1);
    });

    test('Submit Button click without value', () => {
        const mockCallBack = jest.fn();
        const wrapper = mount(<Input name={'test'} onInputSubmit={mockCallBack} />);
        const submitButton = wrapper.find('.submitButton-0-2-2');
        const input = wrapper.find('.input-0-2-1');
        submitButton.simulate('click');
        expect(mockCallBack.mock.calls.length).toEqual(0);
    });


    test('Submit via enter click', () => {
        const mockCallBack = jest.fn();
        const wrapper = mount(<Input name={'test'} onInputSubmit={mockCallBack} />);
        const input = wrapper.find('.input-0-2-1');
        input.simulate('focus');
        input.simulate('change', { target: { value: 'Changed' } });
        input.simulate('keyDown', {key: 'Enter'})
        expect(mockCallBack.mock.calls.length).toEqual(1);
    })
})
