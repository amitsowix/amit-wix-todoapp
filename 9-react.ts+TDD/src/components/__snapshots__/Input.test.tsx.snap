// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input render should render the component correctly 1`] = `
<div
  className="inputContainer-0-2-3"
>
  <input
    autoFocus={true}
    className="input-0-2-1"
    name=""
    onChange={[Function]}
    onKeyDown={[Function]}
    onSubmit={[Function]}
    value=""
  />
  <div
    className="submitButton-0-2-2"
    data-hook="submit-button"
    onClick={[Function]}
  >
    GO
  </div>
</div>
`;
